---
layout: page
title: Reading
permalink: /reading/
---

This page is a reading diary, inspired by [#365papers](https://twitter.com/hashtag/365papers?src=hash) (or 
[#230papers](https://twitter.com/hashtag/230papers?src=hash) realistically).

<div id="papers">

<div style="margin: 0 0 1em 0">

  <p>
    <input class="search" placeholder="Search" />

    <button class="sort" data-sort="author">
      Sort by first author
    </button>

    <button class="sort" data-sort="year">
      Sort by year
    </button>

    <button class="sort" data-sort="read">
      Sort by date read
    </button>

    <input id="abstracts" type="checkbox" name="abstracts" onclick="abstracts();">
    Show abstracts
  </p>

</div>

  <ul class="list">

    <!--
    <li>
      <span class="citation">
        <span class="author">@</span>
        (<span class="year">@</span>)
      </span>
      <a href="@">
        <span class="title">@.</span>
      </a>
      <span class="journal">@.</span>
      <span class="meta">
        <span class="read">2017-02-@</span> 
        <span class="tags"></span>
        <span class="hidden_tags">@</span>
      </span>
      <blockquote class="abstract display_none">@</blockquote>
    </li>
    -->

    <li>
      <span class="citation">
        <span class="author">Templeton et al.</span>
        (<span class="year">1992</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/1385266">
        <span class="title">A cladistic analysis of phenotypic associations with haplotypes inferred from restriction endonuclease mapping and DNA sequence data. 
III. Cladogram estimation.</span>
      </a>
      <span class="journal">Genetics.</span>
      <span class="meta">
        <span class="read">2017-02-12</span> 
        <span class="tags"></span>
        <span class="hidden_tags">drosophila melanogaster; phylogeny; recombination; methods</span>
      </span>
      <blockquote class="abstract display_none">We previously developed a cladistic approach to identify subsets of haplotypes defined by restriction endonuclease 
mapping or DNA sequencing that are associated with significant phenotypic deviations. Our approach was limited to segments of DNA in which little recombination 
occurs. In such cases, a cladogram can be constructed from the restriction site or sequence data that represents the evolutionary steps that interrelate the observed 
haplotypes. The cladogram is used to define a nested statistical design to identify mutational steps associated with significant phenotypic deviations. The central 
assumption behind this strategy is that any undetected mutation causing a phenotypic effect is embedded within the same evolutionary history that is represented by 
the cladogram. The power of this approach depends upon the confidence one has in the particular cladogram used to draw inferences. In this paper, we present a 
strategy for estimating the set of cladograms that are consistent with a particular sample of either restriction site or nucleotide sequence data and that includes 
the possibility of recombination. We first evaluate the limits of parsimony in constructing cladograms. Once these limits have been determined, we construct the set 
of parsimonious and nonparsimonious cladograms that is consistent with these limits. Our estimation procedure also identifies haplotypes that are candidates for 
being products of recombination. If recombination is extensive, our algorithm subdivides the DNA region into two or more subsections, each having little or no 
internal recombination. We apply this estimation procedure to three data sets to illustrate varying degrees of cladogram ambiguity and recombination.</blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Mitchell et al.</span>
        (<span class="year">2014</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/24675797">
        <span class="title">Metabolic and target-site mechanisms combine to confer strong DDT resistance in Anopheles gambiae.</span>
      </a>
      <span class="journal">PLoS One.</span>
      <span class="meta">
        <span class="read">2017-02-09</span> 
        <span class="tags"></span>
        <span class="hidden_tags">journal club; anopheles gambiae; anopheles coluzzii; gste2 I114T; ddt; insecticide resistance; metabolic resistance; vgsc; kdr</span>
      </span>
      <blockquote class="abstract display_none">The development of resistance to insecticides has become a classic exemplar of evolution occurring within human time 
scales. In this study we demonstrate how resistance to DDT in the major African malaria vector Anopheles gambiae is a result of both target-site resistance 
mechanisms that have introgressed between incipient species (the M- and S-molecular forms) and allelic variants in a DDT-detoxifying enzyme. Sequencing of the 
detoxification enzyme, Gste2, from DDT resistant and susceptible strains of An. gambiae, revealed a non-synonymous polymorphism (I114T), proximal to the DDT binding 
domain, which segregated with strain phenotype. Recombinant protein expression and DDT metabolism analysis revealed that the proteins from the susceptible strain 
lost activity at higher DDT concentrations, characteristic of substrate inhibition. The effect of I114T on GSTE2 protein structure was explored through X-ray 
crystallography. The amino acid exchange in the DDT-resistant strain introduced a hydroxyl group nearby the hydrophobic DDT-binding region. The exchange does not 
result in structural alterations but is predicted to facilitate local dynamics and enzyme activity. Expression of both wild-type and 114T alleles the allele in 
Drosophila conferred an increase in DDT tolerance. The 114T mutation was significantly associated with DDT resistance in wild caught M-form populations and acts in 
concert with target-site mutations in the voltage gated sodium channel (Vgsc-1575Y and Vgsc-1014F) to confer extreme levels of DDT resistance in wild caught An. 
gambiae.</blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Hamilton et al.</span>
        (<span class="year">2016</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/27994033">
        <span class="title">Extreme mutation bias and high AT content in Plasmodium falciparum.</span>
      </a>
      <span class="journal">Nucleic Acids Res.</span>
      <span class="meta">
        <span class="read">2017-02-08</span> 
        <span class="tags"></span>
        <span class="hidden_tags">plasmodium falciparum; mutation rate; indel; var</span>
      </span>
      <blockquote class="abstract display_none">For reasons that remain unknown, the Plasmodium falciparum genome has an exceptionally high AT content compared to 
other Plasmodium species and eukaryotes in general - nearly 80% in coding regions and approaching 90% in non-coding regions. Here, we examine how this phenomenon 
relates to genome-wide patterns of de novo mutation. Mutation accumulation experiments were performed by sequential cloning of six P. falciparum isolates growing in 
human erythrocytes in vitro for 4 years, with 279 clones sampled for whole genome sequencing at different time points. Genome sequence analysis of these samples 
revealed a significant excess of G:C to A:T transitions compared to other types of nucleotide substitution, which would naturally cause AT content to equilibrate 
close to the level seen across the P. falciparum reference genome (80.6% AT). These data also uncover an extremely high rate of small indel mutation relative to 
other species, primarily associated with repetitive AT-rich sequences, in addition to larger-scale structural rearrangements focused in antigen-coding var genes. In 
conclusion, high AT content in P. falciparum is driven by a systematic mutational bias and ultimately leads to an unusual level of microstructural plasticity, 
raising the question of whether this contributes to adaptive evolution.</blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Ranson et al.</span>
        (<span class="year">2001</span>)
      </span>
      <a href="http://www.ncbi.nlm.nih.gov/pubmed/11583575">
        <span class="title">Identification of a novel class of insect glutathione S-transferases involved in resistance to DDT in the malaria vector Anopheles 
gambiae.</span>
      </a>
      <span class="journal">Biochem J.</span>
      <span class="meta">
        <span class="read">2017-02-08</span> 
        <span class="tags"></span>
        <span class="hidden_tags">journal club; anopheles gambiae; insecticide resistance; metabolic resistance; gste; ddt</span>
      </span>
      <blockquote class="abstract display_none">The sequence and cytological location of five Anopheles gambiae glutathione S-transferase (GST) genes are described. 
Three of these genes, aggst1-8, aggst1-9 and aggst1-10, belong to the insect class I family and are located on chromosome 2R, in close proximity to previously 
described members of this gene family. The remaining two genes, aggst3-1 and aggst3-2, have a low sequence similarity to either of the two previously recognized 
classes of insect GSTs and this prompted a re-evaluation of the classification of insect GST enzymes. We provide evidence for seven possible classes of insect 
protein with GST-like subunits. Four of these contain sequences with significant similarities to mammalian GSTs. The largest novel insect GST class, class III, 
contains functional GST enzymes including two of the A. gambiae GSTs described in this report and GSTs from Drosophila melanogaster, Musca domestica, Manduca sexta 
and Plutella xylostella. The genes encoding the class III GST of A. gambiae map to a region of the genome on chromosome 3R that contains a major DDT 
[1,1,1-trichloro-2,2-bis-(p-chlorophenyl)ethane] resistance gene, suggesting that this gene family is involved in GST-based resistance in this important malaria 
vector. In further support of their role in resistance, we show that the mRNA levels of aggst3-2 are approx. 5-fold higher in a DDT resistant strain than in the 
susceptible strain and demonstrate that recombinant AgGST3-2 has very high DDT dehydrochlorinase activity.</blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Lee et al.</span>
        (<span class="year">2013</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3856788/">
        <span class="title">Spatiotemporal dynamics of gene flow and hybrid fitness between the M and S forms of the malaria mosquito, Anopheles gambiae.</span>
      </a>
      <span class="journal">Proc Natl Acad Sci USA.</span>
      <span class="meta">
        <span class="read">2017-02-06</span> 
        <span class="tags"></span>
        <span class="hidden_tags">journal club; anopheles gambiae; anopheles coluzzii; speciation; mali; goundry; guinea-bissau</span>
      </span>
      <blockquote class="abstract display_none"> The M and S forms of 
Anopheles gambiae have been the focus of intense study by malaria researchers 
and evolutionary biologists interested in ecological speciation. Divergence 
occurs at three discrete islands in genomes that are otherwise nearly 
identical. An “islands of speciation” model proposes that diverged regions 
contain genes that are maintained by selection in the face of gene flow. An 
alternative “incidental island” model maintains that gene flow between M and S 
is effectively zero and that divergence islands are unrelated to speciation. A 
“divergence island SNP” assay was used to explore the spatial and temporal 
distributions of hybrid genotypes. Results revealed that hybrid individuals 
occur at frequencies ranging between 5% and 97% in every population examined. 
A temporal analysis revealed that assortative mating is unstable and 
periodically breaks down, resulting in extensive hybridization. Results 
suggest that hybrids suffer a fitness disadvantage, but at least some hybrid 
genotypes are viable. Stable introgression of the 2L speciation island 
occurred at one site following a hybridization event.
      </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Prapanthadara et al.</span>
        (<span class="year">1995</span>)
      </span>
      <a 
href="https://www.cambridge.org/core/journals/bulletin-of-entomological-research/article/ddt-resistance-in-anopheles-gambiae-diptera-culicidae-from-zanzibar-tanzania-based-on-increased-ddt-dehydrochlorinase-activity-of-glutathione-s-transferases/BC6B86556732E55E42A44AC9347CDDCE">
        <span class="title">DDT-resistance in Anopheles gambiae (Diptera: 
Culicidae) from Zanzibar, Tanzania, based on increased DDT-dehydrochlorinase 
activity of glutathione S-transferases</span>
      </a>
      <span class="journal">Bulletin of Entomological Research.</span>
      <span class="meta">
        <span class="read">2017-01-31</span> 
        <span class="tags"></span>
        <span class="hidden_tags">anopheles gambiae; insecticide resistance; DDT; metabolic resistance; GST</span>
      </span>
      <blockquote class="abstract display_none"> DDT-resistant Anopheles 
gambiae Giles from Zanzibar, Tanzania, had increased levels of 
DDT-dehydrochlorination compared to a DDT-susceptible strain. Glutathione 
S-transferases (GSTs) are responsible for conversion of DDT to DDE in both the 
susceptible and resistant strains. Sequential column chromatography, including 
Q-Sepharose, S-hexylglutathione agarose, hydroxylapatite and phenyl Sepharose, 
allowed the partial purification of seven GSTs. All seven GSTs possessed 
different degrees of DDTase activity. There was an eight-fold increase in 
total DDTase activity in the resistant compared to the susceptible enzymes. 
Characterization with three substrates, 1-chloro-2,4-dinitrobenzene (CDNB), 
1,2-dichloro-4-nitrobenzene (DCNB) and DDT, revealed the different substrate 
specificity for each isolated GST indicating different isoenzymes. GST Va 
possessed 60% of total DDTase activity suggesting that it contributed most to 
DDT-metabolism in this insect species. The DDTase activity of the GSTs in both 
strains of A. gambiae were found to be correlated with the GST activities 
toward DCNB. Preliminary studies on DDT-resistant and susceptible A. gambiae 
showed that both DDT-resistance and the increased levels of GST activity were 
stage specific which suggested that different GSTs may be involved in 
DDT-resistance in adults and larvae of A. gambiae.
      </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Ranson et al.</span>
        (<span class="year">2000</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/11029667">
        <span class="title">Identification of a point mutation in the 
voltage-gated sodium channel gene of Kenyan Anopheles gambiae associated with 
resistance to DDT and pyrethroids.</span>
      </a>
      <span class="journal">Insect Mol Biol.</span>
      <span class="meta">
        <span class="read">2017-01-31</span> 
        <span class="tags"></span>
        <span class="hidden_tags">journal club; anopheles gambiae; insecticide 
resistance; target-site resistance; DDT; pyrethroids; kdr; vgsc</span>
      </span>
      <blockquote class="abstract display_none"> A field trial of 
permethrin-impregnated bednets and curtains was initiated in Western Kenya in 
1990, and a strain of Anopheles gambiae showing reduced susceptibility to 
permethrin was colonized from this site in 1992. A leucine-phenylalanine 
substitution at position 1014 of the voltage-gated sodium channel is 
associated with resistance to permethrin and DDT in many insect species, 
including Anopheles gambiae from West Africa. We cloned and sequenced a 
partial sodium channel cDNA from the Kenyan permethrin-resistant strain and we 
identified an alternative substitution (leucine to serine) at the same 
position, which is linked to the inheritance of permethrin resistance in the 
F(2) progeny of genetic crosses between susceptible and resistant individuals. 
The diagnostic polymerase chain reaction (PCR) developed by Martinez-Torres et 
al. [(1998) Insect Mol Biol 7: 179-184] to detect kdr alleles in field 
populations of An. gambiae will not detect the Kenyan allele and hence 
reliance on this assay may lead to an underestimate of the prevalence of 
pyrethroid resistance in this species. We adapted the diagnostic PCR to detect 
the leucine-serine mutation and with this diagnostic we were able to 
demonstrate that this kdr allele was present in individuals collected from the 
Kenyan trial site in 1986, prior to the introduction of pyrethroid-impregnated 
bednets. The An. gambiae sodium channel was physically mapped to chromosome 
2L, division 20C. This position corresponds to the location of a major 
quantitative trait locus determining resistance to permethrin in the Kenyan 
strain of An. gambiae.
      </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Berget</span>
        (<span class="year">1995</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/7852296">
        <span class="title">Exon Recognition in Vertebrate Splicing.</span>
      </a>
      <span class="journal">J Biol Chem.</span>
      <span class="meta">
        <span class="read">2017-01-27</span> 
        <span class="tags"></span>
        <span class="hidden_tags">splicing; review</span>
      </span>
      <blockquote class="abstract display_none"> The average vertebrate gene consists of multiple small exons (average size, 137 nucleotides) separated by introns 
that are considerably larger(1). Thus, the vertebrate splicing machinery has the task of finding small desired exons amid much longer introns. The splice site 
consensus sequences that drive exon recognition are located at the very termini of introns(2, 3). Despite the discriminatory challenge faced during exon recognition 
in large multiexon premessenger RNAs, vertebrate splice sites are short and poorly conserved. In fact, splice site sequences in mammals are less conserved than their 
yeast counterparts despite the fact that only a minority of genes in Saccharomyces cerevisiae have introns; and those genes that are split by introns usually have 
only a single intron(4, 5). Thus, vertebrate splicing contends with a more complex specificity problem via recognition of less precise consensus sequences. Any 
mechanism for the orchestration of splicing in multiexon vertebrate genes must provide an explanation for this puzzle.
      </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Norris et al.</span>
        (<span class="year">2015</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/25561525">
        <span class="title">Adaptive introgression in an African malaria mosquito coincident with the increased usage of insecticide-treated bed nets.</span>
      </a>
      <span class="journal">Proc Natl Acad Sci USA.</span>
      <span class="meta">
        <span class="read">2017-01-25</span> 
        <span class="tags"></span>
        <span class="hidden_tags">journal club; anopheles gambiae; anopheles coluzzii; vgsc; kdr; insecticide resistance; target-site resistance; mali</span>
      </span>
      <blockquote class="abstract display_none"> Animal species adapt to changes in their environment, including man-made changes such as the introduction of 
insecticides, through selection for advantageous genes already present in populations or newly arisen through mutation. A possible alternative mechanism is the 
acquisition of adaptive genes from related species via a process known as adaptive introgression. Differing levels of insecticide resistance between two African 
malaria vectors, Anopheles coluzzii and Anopheles gambiae, have been attributed to assortative mating between the two species. In a previous study, we reported two 
bouts of hybridization observed in the town of Selinkenyi, Mali in 2002 and 2006. These hybridization events did not appear to be directly associated with 
insecticide-resistance genes. We demonstrate that during a brief breakdown in assortative mating in 2006, A. coluzzii inherited the entire A. gambiae-associated 2L 
divergence island, which includes a suite of insecticide-resistance alleles. In this case, introgression was coincident with the start of a major insecticide-treated 
bed net distribution campaign in Mali. This suggests that insecticide exposure altered the fitness landscape, favoring the survival of A. coluzzii/A. gambiae 
hybrids, and provided selection pressure that swept the 2L divergence island through A. coluzzii populations in Mali. We propose that the work described herein 
presents a unique description of the temporal dynamics of adaptive introgression in an animal species and represents a mechanism for the rapid evolution of 
insecticide resistance in this important vector of human malaria in Africa.
      </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Davies et al.</span>
        (<span class="year">2007</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/17433068">
        <span class="title">A comparative study of voltage-gated sodium channels in the Insecta: implications for pyrethroid resistance in 
Anopheline and other Neopteran species.</span>
      </a>
      <span class="journal">Insect Mol Biol.</span>
      <span class="meta">
        <span class="read">2017-01-16</span> 
        <span class="tags"></span>
        <span class="hidden_tags">anopheles gambiae; insecticide resistance; target-site resistance; vgsc; kdr; tanzania</span>
      </span>
      <blockquote class="abstract display_none"> We report the complete cDNA sequence of the Anopheles gambiae voltage-gated sodium 
channel (VGSC) alpha-subunit isolated from mature adult mosquitoes. The genomic DNA contains 35 deduced exons with a predicted translation 
of <or= 2139 amino acid cDNAs. The transcription of the gene is, however, complex, alternate splicing being evident for at least five 
optional exons (or exon segments) and two sets of mutually exclusive exons. Overall gene organization was also compared with that of other 
VGSCs within the Insecta. Several insecticides used in mosquito control (including DDT and synthetic pyrethroids) target the VGSC. 
Isolation of the sodium channel cDNA for An. gambiae: (1) allows prediction of likely single nucleotide polymorphisms that may arise at 
residue L1014 to cause resistance to insecticides; (2) defines An. gambiae exon usage in key areas of the VGSC protein that are known 
(from previous studies in a range of different pest species) to have roles in altering insecticide susceptibility and in generating 
resistance; and (3) is a critical first step towards development of refined malarial control strategies and of new diagnostics for 
resistance monitoring.
      </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Dong et al.</span>
        (<span class="year">2014</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4484874/">
        <span class="title">Molecular Biology of Insect Sodium Channels and Pyrethroid Resistance.</span>
      </a>
      <span class="journal">Insect Biochem Mol Biol.</span>
      <span class="meta">
        <span class="read">2017-01-11</span> 
        <span class="tags"></span>
        <span class="hidden_tags">insecticide resistance; target-site resistance; vgsc; kdr; anopheles gambiae; drosophila melanogaster; musca domestica; plutella 
xylostella; blatella germanica</span>
      </span>
      <blockquote class="abstract display_none"> Voltage-gated sodium channels are essential for the initiation and propagation of the action potential in neurons 
and other excitable cells. Because of their critical roles in electrical signaling, sodium channels are targets of a variety of naturally occurring and synthetic 
neurotoxins, including several classes of insecticides. This review is intended to provide an update on the molecular biology of insect sodium channels and the 
molecular mechanism of pyrethroid resistance. Although mammalian and insect sodium channels share fundamental topological and functional properties, most insect 
species carry only one sodium channel gene, compared to multiple sodium channel genes found in each mammalian species. Recent studies showed that two 
posttranscriptional mechanisms, alternative splicing and RNA editing, are involved in generating functional diversity of sodium channels in insects. More than 50 
sodium channel mutations have been identified to be responsible for or associated with knockdown resistance (kdr) to pyrethroids in various arthropod pests and 
disease vectors. Elucidation of molecular mechanism of kdr led to the identification of dual receptor sites of pyrethroids on insect sodium channels. Most of the kdr 
mutations appear to be located within or close to the two receptor sites. The accumulating knowledge of insect sodium channels and their interactions with 
insecticides provides a foundation for understanding the neurophysiology of sodium channels in vivo and the development of new and safer insecticides for effective 
control of arthropod pests and human disease vectors.
      </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Ranson and Lissenden</span>
        (<span class="year">2016</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/26826784">
        <span class="title">Insecticide Resistance in African Anopheles Mosquitoes: A Worsening Situation that Needs Urgent Action to Maintain Malaria
Control.</span>
      </a>
      <span class="journal">Trends Parasitol.</span>
      <span class="meta">
        <span class="read">2017-01-10</span> 
        <span class="tags"></span>
        <span class="hidden_tags">insecticide resistance; anopheles gambiae; anopheles funestus</span>
      </span>
      <blockquote class="abstract display_none"> Malaria control is reliant on insecticides to control the mosquito vector. As efforts to control the disease have 
intensified, so has the selection pressure on mosquitoes to develop resistance to these insecticides. The distribution and strength of this resistance has increased 
dramatically in recent years and now threatens the success of control programs. This review provides an update on the current status of resistance to the major 
insecticide classes in African malaria vectors, considers the evidence that this resistance is already compromising malaria control efforts, and looks to the future 
to highlight some of the new insecticide-based tools under development and the challenges in ensuring they are most effectively deployed to manage resistance.
      </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Lynd et al.</span>
        (<span class="year">2010</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/20056691">
        <span class="title">Field, genetic, and modeling approaches show strong positive selection acting upon an insecticide resistance mutation in Anopheles 
gambiae s.s.</span>
      </a>
      <span class="journal">Mol Biol Evol.</span>
      <span class="meta">
        <span class="read">2017-01-10</span> 
        <span class="tags"></span>
        <span class="hidden_tags">anopheles gambiae; anopheles coluzzii; insecticide resistance; target-site resistance; vgsc; kdr; selection</span>
      </span>
      <blockquote class="abstract display_none"> Alleles subject to strong, recent positive selection will be swept toward fixation together with contiguous sections 
of the genome. Whether the genomic signatures of such selection will be readily detectable in outbred wild populations is unclear. In this study, we employ haplotype 
diversity analysis to examine evidence for selective sweeps around knockdown resistance (kdr) mutations associated with resistance to dichlorodiphenyltrichloroethane 
and pyrethroid insecticides in the mosquito Anopheles gambiae. Both kdr mutations have significantly lower haplotype diversity than the wild-type (nonresistant) 
allele, with kdr L1014F showing the most pronounced footprint of selection. We complement these data with a time series of collections showing that the L1014F allele 
has increased in frequency from 0.05 to 0.54 in 5 years, consistent with a maximum likelihood-fitted selection coefficient of 0.16 and a dominance coefficient of 
0.25. Our data show that strong, recent positive selective events, such as those caused by insecticide resistance, can be identified in wild insect populations.
      </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Gatton et al.</span>
        (<span class="year">2013</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/23550770">
        <span class="title">The importance of mosquito behavioural adaptations to malaria control in Africa.</span>
      </a>
      <span class="journal">Evolution.</span>
      <span class="meta">
        <span class="read">2017-01-09</span> 
        <span class="tags"></span>
        <span class="hidden_tags">journal club; anopheles gambiae; insecticide resistance; behavioural resistance</span>
      </span>
      <blockquote class="abstract display_none"> Over the past decade the use of long-lasting insecticidal nets (LLINs), in combination with improved drug therapies, 
indoor residual spraying (IRS), and better health infrastructure, has helped reduce malaria in many African countries for the first time in a generation. However, 
insecticide resistance in the vector is an evolving threat to these gains. We review emerging and historical data on behavioral resistance in response to LLINs and 
IRS. Overall the current literature suggests behavioral and species changes may be emerging, but the data are sparse and, at times unconvincing. However, preliminary 
modeling has demonstrated that behavioral resistance could have significant impacts on the effectiveness of malaria control. We propose seven recommendations to 
improve understanding of resistance in malaria vectors. Determining the public health impact of physiological and behavioral insecticide resistance is an urgent 
priority if we are to maintain the significant gains made in reducing malaria morbidity and mortality.
      </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Eckhoff et al.</span>
        (<span class="year">2016</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/28028208">
        <span class="title">Impact of mosquito gene drive on malaria elimination in a computational model with explicit spatial and temporal dynamics.</span>
      </a>
      <span class="journal">Proc Natl Acad Sci USA.</span>
      <span class="meta">
        <span class="read">2017-01-05</span> 
        <span class="tags"></span>
        <span class="hidden_tags">gene drive; modelling; anopheles gambiae; anopheles arabiensis</span>
      </span>
      <blockquote class="abstract display_none"> The renewed effort to eliminate malaria and permanently remove its tremendous burden highlights questions of what 
combination of tools would be sufficient in various settings and what new tools need to be developed. Gene drive mosquitoes constitute a promising set of tools, with 
multiple different possible approaches including population replacement with introduced genes limiting malaria transmission, driving-Y chromosomes to collapse a 
mosquito population, and gene drive disrupting a fertility gene and thereby achieving population suppression or collapse. Each of these approaches has had recent 
success and advances under laboratory conditions, raising the urgency for understanding how each could be deployed in the real world and the potential impacts of 
each. New analyses are needed as existing models of gene drive primarily focus on nonseasonal or nonspatial dynamics. We use a mechanistic, spatially explicit, 
stochastic, individual-based mathematical model to simulate each gene drive approach in a variety of sub-Saharan African settings. Each approach exhibits a broad 
region of gene construct parameter space with successful elimination of malaria transmission due to the targeted vector species. The introduction of realistic 
seasonality in vector population dynamics facilitates gene drive success compared with nonseasonal analyses. Spatial simulations illustrate constraints on release 
timing, frequency, and spatial density in the most challenging settings for construct success. Within its parameter space for success, each gene drive approach 
provides a tool for malaria elimination unlike anything presently available. Provided potential barriers to success are surmounted, each achieves high efficacy at 
reducing transmission potential and lower delivery requirements in logistically challenged settings. </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Hu et al.</span>
        (<span class="year">2016</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/28008067">
        <span class="title">Mass seasonal bioflows of high-flying insect migrants.</span>
      </a>
      <span class="journal">Science.</span>
      <span class="meta">
        <span class="read">2017-01-04</span> 
        <span class="tags"></span>
        <span class="hidden_tags">migration</span>
      </span>
      <blockquote class="abstract display_none"> Migrating animals have an impact on ecosystems directly via influxes of predators, prey, and competitors 
and indirectly by vectoring nutrients, energy, and pathogens. Although linkages between vertebrate movements and ecosystem processes have been established, the 
effects of mass insect "bioflows" have not been described. We quantified biomass flux over the southern United Kingdom for high-flying (>150 meters) insects and show 
that ~3.5 trillion insects (3200 tons of biomass) migrate above the region annually. These flows are not randomly directed in insects larger than 10 milligrams, 
which exploit seasonally beneficial tailwinds. Large seasonal differences in the southward versus northward transfer of biomass occur in some years, although flows 
were balanced over the 10-year period. Our long-term study reveals a major transport process with implications for ecosystem services, processes, and 
biogeochemistry. </blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Keightley et al.</span>
        (<span class="year">2014</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/24214343">
        <span class="title">Estimation of the spontaneous mutation rate per nucleotide 
site in a Drosophila melanogaster full-sib family.</span>
      </a>
      <span class="journal">Genetics.</span>
      <span class="meta">
        <span class="read">2017-01-03</span> 
        <span class="tags"></span>
        <span class="hidden_tags">mutation rate; drosophila melanogaster</span>
      </span>
      <blockquote class="abstract display_none"> We employed deep genome 
sequencing of two parents and 12 of their offspring to estimate the mutation rate per 
site per generation in a full-sib family of Drosophila melanogaster recently sampled 
from a natural population. Sites that were homozygous for the same allele in the 
parents and heterozygous in one or more offspring were categorized as candidate 
mutations and subjected to detailed analysis. In 1.23 × 10(9) callable sites from 12 
individuals, we confirmed six single nucleotide mutations. We estimated the false 
negative rate in the experiment by generating synthetic mutations using the empirical 
distributions of numbers of nonreference bases at heterozygous sites in the offspring. 
The proportion of synthetic mutations at callable sites that we failed to detect was 
<1%, implying that the false negative rate was extremely low. Our estimate of the 
point mutation rate is 2.8 × 10(-9) (95% confidence interval = 1.0 × 10(-9) - 6.1 × 
10(-9)) per site per generation, which is at the low end of the range of previous 
estimates, and suggests an effective population size for the species of ∼1.4 × 10(6). 
At one site, point mutations were present in two individuals, indicating that there 
had been a premeiotic mutation cluster, although surprisingly one individual had a G→A 
transition and the other a G→T transversion, possibly associated with error-prone 
mismatch repair. We also detected three short deletion mutations and no insertions, 
giving a deletion mutation rate of 1.2 × 10(-9) (95% confidence interval = 0.7 × 
10(-9) - 11 × 10(-9)).</blockquote>
    </li>

    <li>
      <span class="citation">
        <span class="author">Keightley et al.</span>
        (<span class="year">2015</span>)
      </span>
      <a href="https://www.ncbi.nlm.nih.gov/pubmed/25371432">
        <span class="title">Estimation of the spontaneous mutation rate in 
Heliconius melpomene.</span>
      </a>
      <span class="journal">Mol Biol Evol.</span>
      <span class="meta">
        <span class="read">2017-01-03</span> 
        <span class="tags"></span>
        <span class="hidden_tags">mutation rate; heliconius melpomene</span>
      </span>
      <blockquote class="abstract display_none">We estimated the spontaneous mutation rate in Heliconius 
melpomene by genome sequencing of a pair of parents and 30 of their offspring, 
based on the ratio of number of de novo heterozygotes to the number of 
callable site-individuals. We detected nine new mutations, each one affecting 
a single site in a single offspring. This yields an estimated mutation rate of 
2.9 × 10(-9) (95% confidence interval, 1.3 × 10(-9)-5.5 × 10(-9)), which is 
similar to recent estimates in Drosophila melanogaster, the only other insect 
species in which the mutation rate has been directly estimated. We infer that 
recent effective population size of H. melpomene is about 2 million, a 
substantially lower value than its census size, suggesting a role for natural 
selection reducing diversity. We estimate that H. melpomene diverged from its 
Müllerian comimic H. erato about 6 Ma, a somewhat later date than estimates 
based on a local molecular clock.</blockquote>
    </li>

  </ul>

</div>

<style type="text/css">
.citation {
  font-weight: bold;
}
.meta {
/*  font-style: italic;*/
  color: #777;
}
.display_block {
  display: block;
}
.display_none {
  display: none;
}
.hidden_tags {
  display: none;
}
</style>

<script type="text/javascript" src="/assets/list.min.js"></script>

<script type="text/javascript">

var options = {
  valueNames: [ 'author', 'year', 'title', 'abstract', 'read', 'tags', 'hidden_tags' ]
};

var userList = new List('papers', options);

function abstracts() {
  if (document.getElementById('abstracts').checked) {
    var cls = "abstract display_block";
  } else {
    var cls = "abstract display_none";
  }
  var nodes = document.querySelectorAll(".abstract");
  for (var i=0; i < nodes.length; i++) {
    nodes[i].className = cls;
  }  
}
</script>

This page uses [List.js](https://github.com/javve/list.js).
