<!doctype html>
<html lang="en">

	<head>
		<meta charset="utf-8">

		<title>Tracking the spread of insecticide resistance in Anopheles gambiae populations - 6 June 2019 - Alistair Miles</title>

		<meta name="description" content="Tracking the spread of insecticide resistance in Anopheles gambiae populations - 6 June 2019">
		<meta name="author" content="Alistair Miles">

		<meta name="apple-mobile-web-app-capable" content="yes">
		<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">

		<meta name="viewport" content="width=device-width, initial-scale=1.0">

		<link rel="stylesheet" href="css/reset.css">
		<link rel="stylesheet" href="css/reveal.css">
		<link rel="stylesheet" href="css/theme/white.css" id="theme">

		<!-- Theme used for syntax highlighting of code -->
		<link rel="stylesheet" href="lib/css/monokai.css">

		<!-- Printing and PDF exports -->
		<script>
			var link = document.createElement( 'link' );
			link.rel = 'stylesheet';
			link.type = 'text/css';
			link.href = window.location.search.match( /print-pdf/gi ) ? 'css/print/pdf.css' : 'css/print/paper.css';
			document.getElementsByTagName( 'head' )[0].appendChild( link );
		</script>

		<!--[if lt IE 9]>
		<script src="lib/js/html5shiv.js"></script>
		<![endif]-->

		<style type="text/css">
			/* turn off all caps in titles */
			.reveal h1, .reveal h2, .reveal h3, .reveal h4, .reveal h5 {
				text-transform: none;
			}
		</style>

	</head>

	<body>

		<div class="reveal">

			<div class="slides">

				<section>
					<h2>Tracking the spread of insecticide resistance in <em>Anopheles gambiae</em> populations</h2>
					<p>Alistair Miles (<a href="https://twitter.com/alimanfoo">@alimanfoo</a>)</p>
					<p><a href="http://www.malariagen.net">MalariaGEN Resource Centre</a><br/>University of Oxford / Wellcome Sanger Institute</p>
					<p>6 June 2019 - WHO/Global Malaria Programme</p>
				</section>

				<section>

					<section>
						<div data-markdown>
							## Use cases for genomic surveillance of malaria vector populations

							Pyrethroid resistance is widespread in primary African vector species.

							How can surveillance improve insecticide resistance management (IRM)?
						</div>
						<aside class="notes" data-markdown>
							* When talking about genomic surveillance, helpful to have some use cases in mind.
							* Many possible use cases for genomic surveillance, but I'd just like to mention a couple.
							* These uses cases are focused on the problem of insecticide resistance, and how improved surveillance could lead to improved insecticide resistance management.
						</aside>
					</section>

					<section>
						<div data-markdown>
							## Use case (1): Procurement and deployment of next-generation LLINs

							* "Next-generation" LLINs available, e.g.:
							  * Olyset Plus: permethrin + PBO (P450 synergist)
							  * Olyset Duo: permethrin + pyriproxyfen (second insecticide)
							* More expensive than standard LLINs
							* How many to buy?
							* Where to deploy them?
						</div>
					</section>

					<section>
						<div data-markdown>
							## Use case (2): IRS deployment strategy

							* "Next-generation" IRS formulations available, e.g.:
							  * Actellic 300CS: pyrimiphos methyl (organophosphate)
							  * SumiShield 50WG: clothianidin (neonicotinoid)
							  * Fludora Fusion: deltamethrin + clothianidin
							* Preemptive rotation?
							* Geographical mosaic?
							* Is it working?
						</div>
					</section>

					<section>
						<div data-markdown>
							## Use case (3): Cross-border coordination

							* Can countries take decisions in isolation about how to manage insecticide resistance?
							* When and where do decisions need to be coordinated across borders?
						</div>
					</section>

				</section>



				<section>

					<section>
						<div data-markdown>
							## The *Anopheles gambiae* 1000 Genomes Project (Ag1000G)

							* A consortial project using whole-genome sequencing to investigate genetic variation and evolution in natural mosquito populations
							* Create an open access data resource to accelerate research and surveillance
							* [www.malariagen.net/ag1000g](http://www.malariagen.net/ag1000g)
						</div>
						<aside class="notes" data-markdown>
							* The data I'm presenting today were generated by the Anopheles gambiae 1000 Genomes Consortium
							* This is a consortial project using whole-genome sequencing to investigate genetic variation and evolution in natural mosquito populations
							* All of the data from this project, including both the raw sequence data and our curated and filtered variant calls, are released open access
							* As a resource both for the malaria vector research community and for the development of new approaches to molecular surveillance of mosquito populations
						</aside>
					</section>

					<section>
						<h2>Ag1000G Consortium</h2>
						<p><small>
							Wellcome Sanger Institute / University of Oxford / Liverpool School of Tropical Medicine / Sapienza University of Rome / University of California, Riverside / Liverpool John Moores University / Broad Institute / Institut de Recherche pour le Développement / Virginia Tech / KEMRI Wellcome Trust Research Programme / New Mexico State University / Universidade Nova de Lisboa / University of Minnesota / Université d’Abomey–Calavi, Benin / Indiana University / University of Notre Dame / Washington State University / Imperial College / University of Oregon / University of North Carolina at Chapel Hill / University of Montana / Institut Pasteur / Instituto Nacional de Saúde Pública, Guiné-Bissau / Centre International de Recherches Médicales de Franceville, Gabon / Programa Nacional de Controle da Malária, Angola / Institut de Recherche en Sciences de la Santé, Burkina Faso / University of Bamako, Mali / Infectious Diseases Research Collaboration, Uganda / Organisation de Coordination pour la lutte contre les Endémies en Afrique Centrale, Cameroon
						</small></p>
						<aside class="notes" data-markdown>
							* The Ag1000G project has been driven by a consortium of partners from 29 institutions
							* Who have contributed in a variety of different ways, from mosquito sampling to lab work to @@TODO
						</aside>
					</section>

					<section>
						<div data-markdown>
							## Ag1000G sequencing methods

							* Sequence individual mosquitoes collected from the field
							* Use whole-genome Illumina (Hi-Seq) sequencing
							* Deep coverage (~30X)
							* Sequencing performed at and funded by Wellcome Sanger Institute
						</div>
						<aside class="notes" data-markdown>
							* The methodology we've used is to sequence individual mosquitoes collected directly from the field.
							* We then perform whole genome sequencing using Illumina Hi-Seq machines at the Wellcome Sanger Institute.
							* Each mosquito is sequenced to around 30X coverage, which enables us to make high confidence variant calls.
						</aside>
					</section>

					<section>
						<div data-markdown>
							## Ag1000G population sampling

							* Aim for broad geographical coverage
							  * 18 countries, ~1 site per country
							* *An. gambiae*, *An. coluzzii*, *An. arabiensis*
							* Sequence >30 individuals per site per species
							  * Why 30? Statistical power to make inferences about populations (e.g., gene flow).
						</div>
						<aside class="notes" data-markdown>
							* Our primary aim in Ag1000G has been to achieve broad geographical coverage, so that we have representation from as many different gene pools as possible.
							* In total we have sequenced mosquitoes from 18 African countries.
							* In general we only have mosquitoes from a single site and time point from each country.
							* We've aimed to sequence at least 30 individuals from each site and species.
							* The reason for this is that having a sample of 30 individuals per population gives greater statistical power to make inferences about the populations they were sampled from, such as whether there has been gene flow between them.
						</aside>
					</section>

					<section>
						<div data-markdown>
							## Ag1000G data production

							* Raw sequence reads &rarr;
							* Alignment to reference genome &rarr;
							* Variant calling &rarr;
							* Variant filtering and annotation &rarr;
							* Haplotype phasing &rarr;
							* **Curated "analysis-ready" variant calls and haplotypes**
						</div>
					</section>

					<section>
						<div data-markdown>
							## Ag1000G data releases

							* **Phase 1**: 765 mosquitoes; 8 countries; *An. gambiae*, *An. coluzzii*
							  * Data released 2016
							* **Phase 2**: 1,142 mosquitoes; 13 countries; *An. gambiae*, *An. coluzzii*
							  * Data released 2017
							* **Phase 3**: ~4,000 mosquitoes; 18 countries; *An. gambiae*, *An. coluzzii*, *An. arabiensis*
							  * Data in production
						</div>
					</section>

					<section>
						<div data-markdown>
							## Ag1000G further information

							* www.malariagen.net/ag1000g
							* @@TODO pubmed URL
						</div>
						<img src='./20190606-who-geneva/ag1000g-phase1-paper.png' alt="Ag1000G phase 1 paper"/>
					</section>

				</section>

				<section>

					<section>
						<div data-markdown>
							## Pyrethroid target-site resistance

							Spread of "knock-down resistance" (*kdr*) mutations in the voltage-gated sodium channel gene (*Vgsc*).
						</div>
					</section>

					<section>
						<div data-markdown>
							## *kdr* mutations

							* Two known *kdr* mutations in *Vgsc* codon 1014
							* L1014F found throughout West and Central Africa
							* L1014S found throughout East and Central Africa
							* Are these mutations spreading?
							* Where is gene flow occurring?
						</div>
					</section>

					<section>
						<div data-markdown>
							## Inferring *kdr* gene flow

							* Analyse the genetic backgrounds on which *kdr* mutations occur ("*kdr* haplotypes")
							* Use all mutations within the *Vgsc* gene
							  * 1,710 biallelic SNPs (mostly intronic)
							* Same *kdr* haplotype in two different locations:
							  * &rArr; gene flow
						</div>

					</section>

					<section>
						<h2>Inferring <em>kdr</em> gene flow</h2>
						<img src='./20190606-who-geneva/kdr-gene-flow-cartoon.png' alt="kdr gene flow cartoon"/>
					</section>

					<section>
						<div data-markdown>
							## *kdr* haplotypes

							* Analysed data from Ag1000G phase 1 (765 mosquitoes, 8 countries)
							* L1014F - 5 major haplotypes (F1-F5)
							* L1014S - 5 major haplotypes (S1-S5)
						</div>
					</section>

					<section>
						<h2><em>kdr</em> haplotypes</h2>
						<img src='./20190606-who-geneva/kdr-haplotype-map.png' alt="kdr haplotype map"/>
					</section>

					<section>
						<div data-markdown>
							## *kdr* gene flow - further information

							* @@TODO biorxiv URL
						</div>

					</section>

				</section>

				<section>

					<section>
						<div data-markdown>
							## Pyrethroid metabolic resistance

							Spread of copy number variations in cytochrome P450 genes.
						</div>
					</section>

					<section>
						<div data-markdown>
							## Cytochrome P450 genes

							* Known to metabolise pyrethroids
							* Neutralised by PBO synergist in next-gen LLINs
							* Multiple P450 genes in genome, e.g.:
							  * *Cyp6p/aa*
							  * *Cyp9k1*
							* Increased expression &rArr; pyrethroid resistance
							* Increased gene copy number &rArr; increased expression
						</div>
					</section>

					<section>
						<div data-markdown>
							## Detecting copy number variation

							* Analyse depth of coverage to discover CNVs

							@@TODO image
						</div>
					</section>

					<section>
						<div data-markdown>
							## P450 copy number variation

							* Analysed data from Ag1000G phase 2 (1,142 mosquitoes, 13 countries)
							* Gene amplifications at two P450 loci:
							  * *Cyp6p/aa*
							  * *Cyp9k1*
						</div>
					</section>

					<section>
						<img src='./20190606-who-geneva/cyp6p-cnvs.png' alt="cyp6p CNVs"/>
					</section>

					<section>
						<div data-markdown>
							## *Cyp6p/aa* CNV gene flow

							* Dup1 - BFcol (8%), UGgam (58%)
							* Dup7 - BFcol (44%), CIcol (32%), GHcol (5%), GNcol (75%)
							* Dup8 - BFgam (3/92), GNgam (1/40)
							* Dup10 - BFcol (37/75), GHcol (3/55)
							* Dup11 - CIcol (29/71), GHcol (3/55)
							* Dup14 - BFcol (2/75), CIcol (33/71)
							* Dup15 - BFcol (1/75), CIcol (28/71)
						</div>
					</section>

				</section>

				<section>

					<section data-markdown>
						## Discussion
					</section>

					<section data-markdown>
						## Spread of resistance - open questions

						* Geographical origins?
						* Direction and routes of gene flow?
						* Timing?
						* Rate of movement?
					</section>

					<section data-markdown>
						## Next steps

						* Scale up genome sequencing of vector populations
						* Increase geographical coverage
						* Regular (seasonal) sampling
						* Other vector species (e.g., *An. funestus*)
					</section>

				</section>

				<section>

					<section data-markdown>
						## Acknowledgements

						@@TODO Ag1000G Consortium Partners
					</section>

				</section>

				<section>

					<section data-markdown>
						## Extra slides
					</section>

					<section data-markdown>
						## Gene drive use cases

						* Design of gene drive constructs - identification of ultra-conserved regions
						* Modelling gene drive deployment and spread - estimating rates, ranges and routes of migration
						* Monitoring gene drive in the field - is it spreading as expected, is resistance emerging?
					</section>

					<section>
						<img src="./20190606-who-geneva/kdr-haplotype-networks.png"/>
					</section>

					<section>
						<div data-markdown>
							## *Cyp9k1* CNVs

							@@TODO image of CNV spans, give them names
						</div>
					</section>

				</section>


			</div>

		</div>

		<script src="js/reveal.js"></script>

		<script>

			// More info https://github.com/hakimel/reveal.js#configuration
			Reveal.initialize({
				controls: true,
				progress: true,
				center: true,
				hash: true,

				transition: 'slide', // none/fade/slide/convex/concave/zoom

				// More info https://github.com/hakimel/reveal.js#dependencies
				dependencies: [
					{ src: 'plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
					{ src: 'plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
					{ src: 'plugin/highlight/highlight.js', async: true },
					{ src: 'plugin/search/search.js', async: true },
					{ src: 'plugin/zoom-js/zoom.js', async: true },
					{ src: 'plugin/notes/notes.js', async: true }
				]
			});

		</script>

	</body>
</html>
