---
layout: post
title: Extracting data from VCF files
---


*This post gives an introduction to functions for extracting data from [Variant Call Format (VCF)](http://TODO) files and loading into [NumPy](http://TODO) arrays, [pandas](http://TODO) data frames or [HDF5](http://TODO) files for ease of analysis. These functions are available in [scikit-allel](http://TODO) version 1.1 or later. Any feedback or bug reports welcome.*

## Introduction

### Variant Call Format (VCF)

VCF is a widely-used file format for genetic variation data. Here is an example of a small VCF file, based on the example given in the [VCF specification](https://samtools.github.io/hts-specs/VCFv4.3.pdf):


{% highlight python %}
with open('example.vcf', mode='r') as vcf:
    print(vcf.read(-1))
{% endhighlight %}

    ##fileformat=VCFv4.3
    ##reference=file:///seq/references/1000GenomesPilot-NCBI36.fasta
    ##contig=<ID=20,length=62435964,assembly=B36,md5=f126cdf8a6e0c7f379d618ff66beb2da,species="Homo sapiens",taxonomy=x>
    ##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth">
    ##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency">
    ##INFO=<ID=DB,Number=0,Type=Flag,Description="dbSNP membership, build 129">
    ##FILTER=<ID=q10,Description="Quality below 10">
    ##FILTER=<ID=s50,Description="Less than 50% of samples have data">
    ##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
    ##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth">
    #CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	NA00001	NA00002	NA00003
    20	14370	rs6054257	G	A	29	PASS	DP=14;AF=0.5;DB	GT:DP	0/0:1	0/1:8	1/1:5
    20	17330	.	T	A	3	q10	DP=11;AF=0.017	GT:DP	0/0:3	0/1:5	0/0:41
    20	1110696	rs6040355	A	G,T	67	PASS	DP=10;AF=0.333,0.667;DB	GT:DP	0/2:6	1/2:0	2/2:4
    20	1230237	.	T	.	47	PASS	DP=13	GT:DP	0/0:7	0/0:4	0/0:.
    20	1234567	microsat1	GTC	G,GTCT	50	PASS	DP=9	GT:DP	0/1:4	0/2:2	1/1:3


A VCF file begins with a number of meta-information lines, which start with two hash ('##') characters. Then there is a single header line beginning with a single hash ('#') character. After the header line there are data lines, with each data line describing a genetic variant at a particular position relative to the reference genome of whichever species you are studying. Each data line is divided into fields separated by tab characters. There are 9 fixed fields, labelled "CHROM", "POS", "ID", "REF", "ALT", "QUAL", "FILTER", "INFO" and "FORMAT". Following these are fields containing data about samples. 

For example, the first data line in the file above describes a variant on chromosome 20 at position 14370 relative to the B36 assembly of the human genome. The reference allele is 'G' and the alternate allele is 'A', so this is a single nucleotide polymorphism (SNP). In this file there are three fields with data about samples labelled 'NA00001', 'NA00002' and 'NA00003'. The genotype call in the first sample is '0/0', meaning that individual 'NA0001' is homozygous for the reference allele at this position. The genotype call for the second sample is '0/1' (you may need to scroll across to see this), which means that individual 'NA00002' is heterozygous for the reference and alternate alleles at this position.

### NumPy/pandas/HDF5/...

There are a number of software tools that can read VCF files and perform common analyses. However, if your dataset is large and/or you need to do some bespoke analysis, then it can be faster and more convenient to first extract the necessary data from the VCF file and load into a more efficient storage container.

For analysis and plotting of numerical data in Python, it is very convenient to load data into [NumPy arrays](@@TODO). A NumPy array is an in-memory data structure that provides support for fast arithmetic and data manipulation. For analysing tables of data, [pandas DataFrames](@@TODO) provide useful features such as querying, aggregation and joins. When data are too large to fit into main memory, [HDF5 files](@@TODO) and [Zarr arrays](@@TODO) can provide fast on-disk storage and retrieval of numerical arrays. 

[scikit-allel](@@TODO) is a Python package intended to enable exploratory analysis of large-scale genetic variation data. Version 1.1.0 of scikit-allel adds some new functions for extracting data from VCF files and loading the data into NumPy arrays, pandas DataFrames or HDF5 files. Once you have extracted these data, there are many analyses that can be run interactively on a commodity laptop or desktop computer, even with large-scale datasets from population resequencing studies. To give a flavour of what can be done, there are a few previous articles on my blog, including [variant and sample QC](@@TODO), [allele frequency differentiation](@@TODO), [population structure](@@TODO), and [recombination in genetic crosses](@@TODO).

## [`read_vcf()`](@@TODO)

@@TODO


{% highlight python %}
import allel
print(allel.__version__)
{% endhighlight %}

    1.1.0b5



{% highlight python %}
callset = allel.read_vcf('example.vcf')
{% endhighlight %}


{% highlight python %}
sorted(callset.keys())
{% endhighlight %}




    ['calldata/GT',
     'samples',
     'variants/ALT',
     'variants/CHROM',
     'variants/FILTER_PASS',
     'variants/ID',
     'variants/POS',
     'variants/QUAL',
     'variants/REF']




{% highlight python %}
callset['samples']
{% endhighlight %}




    array(['NA00001', 'NA00002', 'NA00003'], dtype=object)




{% highlight python %}
callset['variants/CHROM']
{% endhighlight %}




    array(['20', '20', '20', '20', '20'], dtype=object)




{% highlight python %}
callset['variants/POS']
{% endhighlight %}




    array([  14370,   17330, 1110696, 1230237, 1234567], dtype=int32)




{% highlight python %}
callset['variants/QUAL']
{% endhighlight %}




    array([ 29.,   3.,  67.,  47.,  50.], dtype=float32)




{% highlight python %}
callset['calldata/GT']
{% endhighlight %}




    array([[[0, 0],
            [0, 1],
            [1, 1]],
    
           [[0, 0],
            [0, 1],
            [0, 0]],
    
           [[0, 2],
            [1, 2],
            [2, 2]],
    
           [[0, 0],
            [0, 0],
            [0, 0]],
    
           [[0, 1],
            [0, 2],
            [1, 1]]], dtype=int8)




{% highlight python %}
gt = allel.GenotypeArray(callset['calldata/GT'])
gt
{% endhighlight %}




<div class="allel allel-DisplayAs2D"><span>&lt;GenotypeArray shape=(5, 3, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style="text-align: center">0</th><th style="text-align: center">1</th><th style="text-align: center">2</th></tr></thead><tbody><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">0</th><td style="text-align: center">0/0</td><td style="text-align: center">0/1</td><td style="text-align: center">1/1</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">1</th><td style="text-align: center">0/0</td><td style="text-align: center">0/1</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">2</th><td style="text-align: center">0/2</td><td style="text-align: center">1/2</td><td style="text-align: center">2/2</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">3</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">4</th><td style="text-align: center">0/1</td><td style="text-align: center">0/2</td><td style="text-align: center">1/1</td></tr></tbody></table></div>




{% highlight python %}
gt.is_het()
{% endhighlight %}




    array([[False,  True, False],
           [False,  True, False],
           [ True,  True, False],
           [False, False, False],
           [ True,  True, False]], dtype=bool)




{% highlight python %}
gt.count_het(axis=1)
{% endhighlight %}




    array([1, 1, 2, 0, 2])




{% highlight python %}
ac = gt.count_alleles()
ac
{% endhighlight %}




<div class="allel allel-DisplayAs2D"><span>&lt;AlleleCountsArray shape=(2, 4) dtype=int32&gt;</span><table><thead><tr><th></th><th style="text-align: center">0</th><th style="text-align: center">1</th><th style="text-align: center">2</th><th style="text-align: center">3</th></tr></thead><tbody><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">0</th><td style="text-align: center">9</td><td style="text-align: center">3</td><td style="text-align: center">0</td><td style="text-align: center">0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">1</th><td style="text-align: center">1</td><td style="text-align: center">3</td><td style="text-align: center">3</td><td style="text-align: center">5</td></tr></tbody></table></div>



### Selecting fields

@@TODO


{% highlight python %}
callset = allel.read_vcf('example.vcf', fields=['DP', 'calldata/HQ'])
sorted(callset.keys())
{% endhighlight %}




    ['calldata/HQ', 'variants/DP']




{% highlight python %}
callset['variants/DP']
{% endhighlight %}




    array([14, 11, 10, 13,  9], dtype=int32)




{% highlight python %}
callset['calldata/HQ']
{% endhighlight %}




    array([[[51, 51],
            [51, 51],
            [-1, -1]],
    
           [[58, 50],
            [65,  3],
            [-1, -1]],
    
           [[23, 27],
            [18,  2],
            [-1, -1]],
    
           [[56, 60],
            [51, 51],
            [-1, -1]],
    
           [[-1, -1],
            [-1, -1],
            [-1, -1]]], dtype=int8)




{% highlight python %}
callset = allel.read_vcf('example.vcf', fields='*')
sorted(callset.keys())
{% endhighlight %}




    ['calldata/DP',
     'calldata/GQ',
     'calldata/GT',
     'calldata/HQ',
     'samples',
     'variants/AA',
     'variants/AF',
     'variants/ALT',
     'variants/CHROM',
     'variants/DB',
     'variants/DP',
     'variants/FILTER_PASS',
     'variants/FILTER_q10',
     'variants/FILTER_s50',
     'variants/H2',
     'variants/ID',
     'variants/NS',
     'variants/POS',
     'variants/QUAL',
     'variants/REF',
     'variants/is_snp',
     'variants/numalt',
     'variants/svlen']



### Data type

@@TODO


{% highlight python %}
callset = allel.read_vcf('example.vcf', fields=['DP'])
callset['variants/DP']
{% endhighlight %}




    array([14, 11, 10, 13,  9], dtype=int32)




{% highlight python %}
callset = allel.read_vcf('example.vcf', fields=['DP'], types={'DP': 'int16'})
callset['variants/DP']
{% endhighlight %}




    array([14, 11, 10, 13,  9], dtype=int16)




{% highlight python %}
callset = allel.read_vcf('example.vcf', fields=['DP'], types={'DP': 'float32'})
callset['variants/DP']
{% endhighlight %}




    array([ 14.,  11.,  10.,  13.,   9.], dtype=float32)




{% highlight python %}
callset = allel.read_vcf('example.vcf', fields=['REF'])
callset['variants/REF']
{% endhighlight %}




    array(['G', 'T', 'A', 'T', 'GTC'], dtype=object)




{% highlight python %}
callset = allel.read_vcf('example.vcf', fields=['REF'], types={'REF': 'S3'})
callset['variants/REF']
{% endhighlight %}




    array([b'G', b'T', b'A', b'T', b'GTC'], 
          dtype='|S3')




{% highlight python %}
callset = allel.read_vcf('example.vcf', fields=['REF'], types={'REF': 'S1'})
callset['variants/REF']
{% endhighlight %}




    array([b'G', b'T', b'A', b'T', b'G'], 
          dtype='|S1')



### Number of values

@@TODO


{% highlight python %}
callset = allel.read_vcf('example.vcf', fields=['ALT'])
callset['variants/ALT']
{% endhighlight %}




    array([['A', '', ''],
           ['A', '', ''],
           ['G', 'T', ''],
           ['', '', ''],
           ['G', 'GTCT', '']], dtype=object)




{% highlight python %}
callset = allel.read_vcf('example.vcf', fields=['ALT'], numbers={'ALT': 5})
callset['variants/ALT']
{% endhighlight %}




    array([['A', '', '', '', ''],
           ['A', '', '', '', ''],
           ['G', 'T', '', '', ''],
           ['', '', '', '', ''],
           ['G', 'GTCT', '', '', '']], dtype=object)




{% highlight python %}
callset = allel.read_vcf('example.vcf', fields=['ALT'], numbers={'ALT': 1})
callset['variants/ALT']
{% endhighlight %}




    array(['A', 'A', 'G', '', 'G'], dtype=object)



### Genotype ploidy

@@TODO


{% highlight python %}
with open('example_polyploid.vcf', mode='r') as f:
    print(f.read(-1))
{% endhighlight %}

    ##fileformat=VCFv4.3
    ##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
    #CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	sample1	sample2	sample3
    20	14370	.	G	A	.	.	.	GT	0/0/0/0	0/0/0/1	0/0/1/1
    20	17330	.	T	A,C,G	.	.	.	GT	1/1/2/2	0/1/2/3	3/3/3/3
    



{% highlight python %}
callset = allel.read_vcf('example_polyploid.vcf', fields=['calldata/GT'], numbers={'GT': 4})
callset['calldata/GT']
{% endhighlight %}




    array([[[0, 0, 0, 0],
            [0, 0, 0, 1],
            [0, 0, 1, 1]],
    
           [[1, 1, 2, 2],
            [0, 1, 2, 3],
            [3, 3, 3, 3]]], dtype=int8)




{% highlight python %}
gt = allel.GenotypeArray(callset['calldata/GT'])
gt
{% endhighlight %}




<div class="allel allel-DisplayAs2D"><span>&lt;GenotypeArray shape=(2, 3, 4) dtype=int8&gt;</span><table><thead><tr><th></th><th style="text-align: center">0</th><th style="text-align: center">1</th><th style="text-align: center">2</th></tr></thead><tbody><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">0</th><td style="text-align: center">0/0/0/0</td><td style="text-align: center">0/0/0/1</td><td style="text-align: center">0/0/1/1</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">1</th><td style="text-align: center">1/1/2/2</td><td style="text-align: center">0/1/2/3</td><td style="text-align: center">3/3/3/3</td></tr></tbody></table></div>




{% highlight python %}
gt.is_het()
{% endhighlight %}




    array([[False,  True,  True],
           [ True,  True, False]], dtype=bool)




{% highlight python %}
ac = gt.count_alleles()
ac
{% endhighlight %}




<div class="allel allel-DisplayAs2D"><span>&lt;AlleleCountsArray shape=(2, 4) dtype=int32&gt;</span><table><thead><tr><th></th><th style="text-align: center">0</th><th style="text-align: center">1</th><th style="text-align: center">2</th><th style="text-align: center">3</th></tr></thead><tbody><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">0</th><td style="text-align: center">9</td><td style="text-align: center">3</td><td style="text-align: center">0</td><td style="text-align: center">0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">1</th><td style="text-align: center">1</td><td style="text-align: center">3</td><td style="text-align: center">3</td><td style="text-align: center">5</td></tr></tbody></table></div>



### Selecting a genome region

@@TODO


{% highlight python %}
callset = allel.read_vcf('example.vcf', region='20:1000000-1231000')
callset['variants/CHROM'], callset['variants/POS']
{% endhighlight %}




    (array(['20', '20'], dtype=object), array([1110696, 1230237], dtype=int32))



### Selecting samples

@@TODO


{% highlight python %}
callset = allel.read_vcf('example.vcf', samples=['NA00001', 'NA00003'])
callset['samples']
{% endhighlight %}




    array(['NA00001', 'NA00003'], dtype=object)




{% highlight python %}
allel.GenotypeArray(callset['calldata/GT'])
{% endhighlight %}




<div class="allel allel-DisplayAs2D"><span>&lt;GenotypeArray shape=(5, 2, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style="text-align: center">0</th><th style="text-align: center">1</th></tr></thead><tbody><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">0</th><td style="text-align: center">0/0</td><td style="text-align: center">1/1</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">1</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">2</th><td style="text-align: center">0/2</td><td style="text-align: center">2/2</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">3</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">4</th><td style="text-align: center">0/1</td><td style="text-align: center">1/1</td></tr></tbody></table></div>



## [`vcf_to_npz()`](@@TODO)

@@TODO


{% highlight python %}

{% endhighlight %}

## [`vcf_to_hdf5()`](@@TODO)

@@TODO


{% highlight python %}

{% endhighlight %}

## [`vcf_to_zarr()`](@@TODO)

@@TODO


{% highlight python %}

{% endhighlight %}

## [`vcf_to_dataframe()`](@@TODO)

@@TODO


{% highlight python %}

{% endhighlight %}

## [`vcf_to_csv()`](@@TODO)

@@TODO


{% highlight python %}

{% endhighlight %}

## [`vcf_to_recarray()`](@@TODO)

@@TODO


{% highlight python %}

{% endhighlight %}

## Worked examples

### Human 1000 genomes phase 3

@@TODO


{% highlight python %}

{% endhighlight %}

### Pf3k (Plasmodium falciparum) release 5.1

@@TODO

## Other datasets

### Ag1000G (Anopheles gambiae) phase 1 AR3 release

@@TODO

## Further reading

@@TODO


## Post-script: changes from `vcfnp`

The new functions available in `scikit-allel` supercede a package I previously wrote for extracting data from VCF files called [`vcfnp`](@@TODO). I rewrote this functionality from the ground up and ported the functionality to `scikit-allel` for two main reasons. Firstly, `vcfnp` was slow and so you needed a cluster to parse big VCF files, which is obviously a pain. The new functions in `scikit-allel` should be up to ~40 times faster. Secondly, the `vcfnp` API was somewhat complicated, requiring three separate steps to get data from VCF into an HDF5 file or Zarr store. The new functions in `scikit-allel` hopefully simplify this process, enabling data to be extracted from VCF and loaded into any of a variety of storage containers via a single function call.

If you previously used `vcfnp` here are a few notes on some of the things that have changed.

* No need for separate function calls to extract data from variants and calldata fields, both can be extracted via a single call to `read_vcf()` or any of the `vcf_to_...()` functions described above.
* Data can be extracted from VCF and loaded into HDF5 with a single function call to `vcf_to_hdf5()`; i.e., no need to first extract parts of the data out to .npy files then load into HDF5.
* No need to use a cluster or do any parallelisation, it should be possible to run `vcf_to_hdf5()` or `vcf_to_zarr()` on a whole VCF on a half-decent desktop or laptop computer, although big VCF files might take a couple of hours and require a reasonably large hard disk.
* The default NumPy data type for string fields has changed to use 'object' dtype, which means that strings of any length will be stored automatically (i.e., no need to configure separate dtypes for each string field) and there will be no truncation of long strings.
* Previously in `vcfnp` the genotype calls were extracted into a special field called 'genotype' separate from the 'GT' calldata field if requested. In `scikit-allel` the default behaviour is to parse the 'GT' field as a 3-dimensional integer array and return simply as 'calldata/GT'. If you really want to process the 'GT' field as a string then you can override this by setting the type for 'calldata/GT' to 'S3' or 'object'.



{% highlight python %}

{% endhighlight %}
