---
layout: post
title: A tour of Ag1000G phase2
---


Today we made the first [data release](https://www.malariagen.net/data/ag1000g-phase2-ar1) from phase 2 of the *Anopheles gambiae* 1000 genomes project ([Ag1000G](http://www.malariagen.net/ag1000g)). These data include variant calls and haplotypes for 1,142 wild-caught mosquitoes from 13 African countries, and 234 mosquitoes from 11 lab crosses. In this article I thought I would give a quick tour of the data release, summarizing some of the main features of the data.

The data are available for download from a [public FTP site](ftp://ngs.sanger.ac.uk/production/ag1000g/phase2/AR1/). I have a copy of some of the files downloaded to a directory on my computer, so I'll be loading data from there.


{% highlight python %}
release_dir = 'data/ag1000g/phase2/AR1'
{% endhighlight %}

## Population sampling

Some metadata about the mosquitoes we've sampled is available in the "samples" sub-directory. Let's load metadata for the wild-caught mosquitoes.


{% highlight python %}
import os
import pandas as pd
samples = pd.read_csv(os.path.join(release_dir, 'samples', 'samples.meta.txt'),
                      sep='\t', index_col='ox_code')
samples.head()
{% endhighlight %}




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>src_code</th>
      <th>population</th>
      <th>country</th>
      <th>region</th>
      <th>contributor</th>
      <th>contact</th>
      <th>year</th>
      <th>m_s</th>
      <th>sex</th>
      <th>n_sequences</th>
      <th>mean_coverage</th>
    </tr>
    <tr>
      <th>ox_code</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AA0040-C</th>
      <td>Twifo_Praso__E2</td>
      <td>GHcol</td>
      <td>Ghana</td>
      <td>Twifo_Praso</td>
      <td>David Weetman</td>
      <td>NaN</td>
      <td>2012.0</td>
      <td>M</td>
      <td>F</td>
      <td>95033368</td>
      <td>30.99</td>
    </tr>
    <tr>
      <th>AA0041-C</th>
      <td>Twifo_Praso__H3</td>
      <td>GHcol</td>
      <td>Ghana</td>
      <td>Twifo_Praso</td>
      <td>David Weetman</td>
      <td>NaN</td>
      <td>2012.0</td>
      <td>M</td>
      <td>F</td>
      <td>95843804</td>
      <td>31.70</td>
    </tr>
    <tr>
      <th>AA0042-C</th>
      <td>Takoradi_C7</td>
      <td>GHcol</td>
      <td>Ghana</td>
      <td>Takoradi</td>
      <td>David Weetman</td>
      <td>NaN</td>
      <td>2012.0</td>
      <td>M</td>
      <td>F</td>
      <td>107420666</td>
      <td>35.65</td>
    </tr>
    <tr>
      <th>AA0043-C</th>
      <td>Takoradi_H8</td>
      <td>GHcol</td>
      <td>Ghana</td>
      <td>Takoradi</td>
      <td>David Weetman</td>
      <td>NaN</td>
      <td>2012.0</td>
      <td>M</td>
      <td>F</td>
      <td>95993752</td>
      <td>29.46</td>
    </tr>
    <tr>
      <th>AA0044-C</th>
      <td>Takoradi_D10</td>
      <td>GHcol</td>
      <td>Ghana</td>
      <td>Takoradi</td>
      <td>David Weetman</td>
      <td>NaN</td>
      <td>2012.0</td>
      <td>M</td>
      <td>F</td>
      <td>103044262</td>
      <td>33.67</td>
    </tr>
  </tbody>
</table>
</div>



The "ox_code" column is the main identifier we use for each mosquito in our analyses. Just to confirm how many individuals:


{% highlight python %}
len(samples)
{% endhighlight %}




    1142



Here's a breakdown of number of mosquitoes by country:


{% highlight python %}
samples.country.value_counts()
{% endhighlight %}




    Cameroon             297
    Burkina Faso         167
    Uganda               112
    Guinea-Bissau         91
    Angola                78
    Cote d'Ivoire         71
    Gabon                 69
    Ghana                 67
    Gambia, The           65
    Kenya                 48
    Guinea                44
    France                24
    Equatorial Guinea      9
    Name: country, dtype: int64



The mosquitoes from "France" were collected on Mayotte Island, and the mosquitoes from Equatorial Guinea were collected on Bioko Island.

### Population definitions

Ag1000G phase 2 includes mosquitoes from two species, *An. gambiae* and *An. coluzzii*. It also includes mosquitoes from populations which are hard to assign unambiguously to either *An. gambiae* or *An. coluzzii* because of some apparent mixed ancestry. To aid with downstream analyses we have assigned each mosquito to one of 16 populations, based on country of origin and species. Here's a breakdown of number of mosquitoes by population:


{% highlight python %}
samples.population.value_counts()
{% endhighlight %}




    CMgam    297
    UGgam    112
    BFgam     92
    GW        91
    AOcol     78
    BFcol     75
    CIcol     71
    GAgam     69
    GM        65
    GHcol     55
    KE        48
    GNgam     40
    FRgam     24
    GHgam     12
    GQgam      9
    GNcol      4
    Name: population, dtype: int64



Each population identifier is formed by concatenating the two letter country code (e.g., "CM" for Cameroon) with an abbreviation for the species ("gam" means *An. gambiae*, "col" means *An. coluzzii*). There are three populations (GW, GM, KE) where we have not divided by species because of mixed ancestry. For all other populations, the assignment of species for each individual was based on the results of the conventional PCR-based molecular tests.

### Lab crosses

There are also a total of 11 lab crosses included in this release. Each cross comprises 2 parents and up to 20 progeny. 


{% highlight python %}
crosses = pd.read_csv(os.path.join(release_dir, 'samples', 'cross.samples.meta.txt'),
                      sep='\t', usecols=range(1, 9), index_col='ox_code')
crosses.head()
{% endhighlight %}




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cross</th>
      <th>role</th>
      <th>n_reads</th>
      <th>median_cov</th>
      <th>mean_cov</th>
      <th>sex</th>
      <th>colony_id</th>
    </tr>
    <tr>
      <th>ox_code</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AD0142-C</th>
      <td>18-5</td>
      <td>parent</td>
      <td>60486753</td>
      <td>26</td>
      <td>25.824447</td>
      <td>F</td>
      <td>Ghana</td>
    </tr>
    <tr>
      <th>AD0143-C</th>
      <td>18-5</td>
      <td>parent</td>
      <td>58509103</td>
      <td>19</td>
      <td>18.800118</td>
      <td>M</td>
      <td>Kisumu/G3</td>
    </tr>
    <tr>
      <th>AD0146-C</th>
      <td>18-5</td>
      <td>progeny</td>
      <td>101612499</td>
      <td>44</td>
      <td>43.494594</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>AD0147-C</th>
      <td>18-5</td>
      <td>progeny</td>
      <td>50710020</td>
      <td>16</td>
      <td>16.284487</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>AD0148-C</th>
      <td>18-5</td>
      <td>progeny</td>
      <td>59023991</td>
      <td>19</td>
      <td>18.978021</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>



Each cross has been given an identifier like "18-5", these are just arbitrary identifiers and don't mean anything. Here's a breakdown of number of individuals by cross:


{% highlight python %}
crosses.cross.value_counts()
{% endhighlight %}




    37-3    22
    46-9    22
    47-6    22
    80-2    22
    45-1    22
    36-9    22
    29-2    22
    18-5    22
    73-2    21
    78-2    21
    42-4    16
    Name: cross, dtype: int64



The parents of the crosses came from various commonly used lab colonies, e.g., "Mali" or "Pimperena". Because of the way the crosses were performed, in some cases we could not be completely certain of the parent colony, and these are labelled as ambiguous, e.g.., "Kisumu/G3". Here's a count of which colonies were used for the parents:


{% highlight python %}
crosses[crosses.role == 'parent'].colony_id.value_counts()
{% endhighlight %}




    Mali            6
    Kisumu          5
    Ghana           4
    Kisumu/Ghana    2
    Akron           2
    Pimperena       2
    Kisumu/G3       1
    Name: colony_id, dtype: int64




{% highlight python %}

{% endhighlight %}
